jobrunner/run.py

```diff
@@ -195,28 +155,142 @@ def handle_running_job(job):
             cleanup_job(job)
 
 
-def handle_running_job_api(job, api):
+def handle_active_jobs_api(api, active_jobs):
+    for job in active_jobs:
+        # `set_log_context` ensures that all log messages triggered anywhere
+        # further down the stack will have `job` set on them
+        with set_log_context(job=job):
+            try:
+                handle_job_api(job, api)
+            except Exception:
+                mark_job_as_failed(job, f"Internal error")
+                # Do not clean up, as we may want to debug
+                #
+                # Raising will kill the main loop, by design. The service manager
+                # will restart, and this job will be ignored when it does, as
+                # it has failed. If we have an internal error, a full restart
+                # might recover better.
+                raise
+
+
+def handle_job_api(job, api):
+    """Handle an active job.
+
+    This contains the main state machine logic for a job. For the most part,
+    state transitions follow the same logic, which is abstracted. Some
+    transitions require special logic, mainly the initial and final states, as
+    well as supporting cancellation.
+    """
+    assert job.state in (State.PENDING, State.RUNNING)
+    code = None
+    message = None
+
+    definition = job_to_job_definition(job)
+    old_status = api.get_status(definition)
+
     if job.cancelled:
-        log.info("Cancellation requested, killing job")
-        api.terminate(job_to_job_definition(job))
+        # cancelled is driven by user request, so is handled explicitly first
+        # regardless of executor state.
+        api.terminate(definition)
```

Comment:
Since the API returns a state on every call, I think we should at least log it rather than throwing it away. Maybe a decorator is in order which logs every call and the response. That could go in a separate card rather than going to that effort here.

Reply:
a state transition log is a good idea.

Reply:
https://linear.app/opensafely/issue/OS-68/add-state-transition-log